src/evict/evict_lru.c

```diff
@@ -1539,44 +1541,44 @@ __evict_walk_file(WT_SESSION_IMPL *session,
 		bytes_per_slot = 1 + cache_inuse / total_slots;
 		target_pages_clean = (uint32_t)(
 		    (btree_inuse + bytes_per_slot / 2) / bytes_per_slot);
-	} else
-		target_pages_clean = 0;
+	}
 
 	if (F_ISSET(cache, WT_CACHE_EVICT_DIRTY)) {
 		btree_inuse = __wt_btree_dirty_leaf_inuse(session);
 		cache_inuse = __wt_cache_dirty_leaf_inuse(cache);
 		bytes_per_slot = 1 + cache_inuse / total_slots;
 		target_pages_dirty = (uint32_t)(
 		    (btree_inuse + bytes_per_slot / 2) / bytes_per_slot);
-	} else
-		target_pages_dirty = 0;
+	}
 
 	target_pages = WT_MAX(target_pages_clean, target_pages_dirty);
 
-	if (target_pages == 0) {
-		/*
-		 * Randomly walk trees with a tiny fraction of the cache in
-		 * case there are so many trees that none of them use enough of
-		 * the cache to be allocated slots.  Walk small trees 1% of the
-		 * time.
-		 */
-		if (__wt_random(&session->rnd) > UINT32_MAX / 100)
-			return (0);
+	/*
+	 * Randomly walk trees with a small fraction of the cache in case there
+	 * are so many trees that none of them use enough of the cache to be
+	 * allocated slots.  Make the chance of walking a tree equal to the
+	 * chance that a random byte in cache belongs to the tree.
+	 */
+	if (target_pages == 0 && (cache_inuse == 0 ||
+	    __wt_random64(&session->rnd) % cache_inuse > btree_inuse))
```

Comment:
@agorrod, I'll push a fix for the dirty / clean calculation shortly.

Here is a slightly different way to slice that scenario:

| Tree type | Number of trees | Chance of selection | Pages per pass
|---|---|---|---|
| Type A | 800 | 0.001% | 0.08 |
| Type B | 100 | 0.1% | 1 |
| Type C | 50 | 0.5% | 2.5 |
| Type D | 50 | 1.3% | 6.5 |

Pages per pass = Number of trees * Chance of selection * 10 (pages per tree is constant here)

I think you are saying that in this case where all trees are "small", we should expect to find more than 10 pages after walking through all the trees in the cache.  Is that right?

The actual proportions there look right to me -- i.e., we are putting more effort into the bigger trees.  If we were to bias the probability based on the total number of *slots* (rather than trees), would that address this concern?